### PR TITLE
doc: posix: option groups: abbreviate c option group, add missing thread safe function table

### PR DESCRIPTION
**Pull request by cfriedt**
* reduce verbosity in `POSIX_C_LANG_SUPPORT` and link to details, language support page
* add missing `POSIX_C_LANG_JUMP` section
* add missing `POSIX_C_LANG_MATH` section and link to details
* add missing `_POSIX_THREAD_SAFE_FUNCTIONS`
* add clarification on POSIX API features in other C libraries

### CODE REVIEW

**Review by ycsin — DISMISSED** [review]

File: doc/services/portability/posix/option_groups/posix-c-lang-support.rst [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+The ``POSIX_C_LANG_SUPPORT`` option group is supplied by the ISO C Library.
+
+Zephyr supports most ISO C features pertaining to the OS. These features are cooperatively
+maintained by the C Library and POSIX API maintainers.
+
+For more information on developing Zephry applications in C, or other programming languages,
+please refer to :ref:`details<language_support>`.
+
+.. csv-table:: POSIX_C_LANG_SUPPORT
```
Comment (kartben) [inline doc/services/portability/posix/option_groups/posix-c-lang-support.rst:19]: What's the rationale for the functions listed as "no" below? Like, `asctime()` is available in Newline and Picolibc, and even used in the gettimeofday sample?

**Review by kartben — DISMISSED** [review]

File: doc/services/portability/posix/option_groups/posix-c-lang-support.rst (hunk 1 above)
Comment (cfriedt) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst:19]: That's a good question. I assumed the C Library maintainer would be updating this list periodically as well. Hence, the PR.

File: doc/services/portability/posix/option_groups/posix-c-lang-support.rst [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+POSIX_C_LANG_SUPPORT
+====================
+
+The ``POSIX_C_LANG_SUPPORT`` option group is supplied by the ISO C Library.
+
+Zephyr supports most ISO C features pertaining to the OS. These features are cooperatively
+maintained by the C Library and POSIX API maintainers.
+
+For more information on developing Zephry applications in C, or other programming languages,
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst]: Fixed!

File: doc/services/portability/posix/option_groups/posix-c-lang-support.rst (hunk 1 above)
Comment (cfriedt) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst:19]: I guess we can only declare functions as supported if they are also supported in the minimal libc, and it looks like `asctime()` is unavailable there. I would guess that's the case for most other unimplemented functions.

However, picolibc, and newlib can supply most (if not all) of these functions, and picolibc is IIRC the default now (but only if either the Zephyr SDK is available or picolibc is available as a module).

Comment (cfriedt) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst:37]: Added some paragraphs explaining the significance of this table.

**Comment by cfriedt** [comment]
* removed RTEMS GSoC item from timeline / jpg

**Comment by cfriedt** [comment]
* moved unrelated changes to separate prs

**Review by ycsin — DISMISSED** [review]
LGTM - https://builds.zephyrproject.io/zephyr/pr/66868/docs/services/portability/posix/index.html

**Comment by cfriedt** [comment]
ping @stephanosio @kartben @carlescufi

Please approve / merge. Thanks!

File: doc/services/portability/posix/option_groups/posix-c-lang-support.rst (hunk 1 above)
Comment (kartben) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst:34]: > I honestly wonder if maybe this list should be moved to [Minimal libc](https://docs.zephyrproject.org/latest/develop/languages/c/minimal_libc.html) since it appears to mainly reflect completion status of that library.

+1. Keeping the table is making the whole page unnecessarily verbose IMO, i.e. I don't think it's the right place for delving into Zephyr history, and it should rather focus on current/default behavior. In short: I would remove the table altogether, and remove L14-L29 to replace it with something along the lines of the second note ("When using Newlib, Picolibc (default C library as of Zephyr 3.5), or other C libraries conforming to the ISO C Standard, the entire
   ``POSIX_C_LANG_SUPPORT`` Option Group is considered supported.

Comment (cfriedt) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst]: Docs are building fine on my end now - I updated to include a couple of things that I was surprised were missing. 

`POSIX_C_LANG_JUMP` (already supported, + link to details)
`POSIX_C_LANG_MATH` (+ link to details)

Also made a small comment that should clarify that the C library is also a perfectly valid source for conforment implementations of POSIX API (of course). It depends on the C library and the implementation, as to whether it's going to work with Zephyr, of course. E.g. should pass testsuite.

Comment (cfriedt) [reply doc/services/portability/posix/option_groups/posix-c-lang-support.rst]: @kartben , @keith-packard - PTAL

File: doc/services/portability/posix/option_groups/index.rst [hunk 3]
```diff
@@ -102,123 +102,54 @@ This table lists service support status in Zephyr:
...
+
+    setjmp(), yes
+    longjmp(), yes
+
+.. _posix_option_group_c_lang_math:
+
+POSIX_C_LANG_MATH
+=================
+
+The ``POSIX_C_LANG_MATH`` Option Group is included in the ISO C standard.
```
Comment (keith-packard) [inline doc/services/portability/posix/option_groups/index.rst:128]: POSIX has extensions to the ISO C specification, including numerous clarifications of math function behavior. Does Zephyr require conformance to POSIX specs or to the ISO C spec for these functions?

Comment (cfriedt) [reply doc/services/portability/posix/option_groups/index.rst:128]: Excellent question! The official answer is here
https://pubs.opengroup.org/onlinepubs/9699919799/help/codes.html#CX

TLDR - the ISO spec should be considered the correct one.

Comment (keith-packard) [reply doc/services/portability/posix/option_groups/index.rst:128]: This doesn't answer my question though -- for ISO C APIs where the POSIX specification places additional requirements on the implementation, can applications rely on the POSIX spec or must they only rely on the ISO C spec? That would require auditing all uses of ISO C APIs within any POSIX code brought into the Zephyr environment to ensure they do not rely on any POSIX extensions to the ISO C specification.

Given that POSIX specifications are not supposed to conflict with ISO C specifications, but only to refine them, it seems like Zephyr should require a POSIX conforming C library when the application requests POSIX functionality? Or at least make it clear to the application that they are operating outside known behavior when mixing non-POSIX implementations of C library functions with POSIX functionality.

Comment (cfriedt) [reply doc/services/portability/posix/option_groups/index.rst:128]: > This doesn't answer my question though -- for ISO C APIs where the POSIX specification places additional requirements on the implementation, can applications rely on the POSIX spec or must they only rely on the ISO C spec?

I think it means use the iso spec and consider the POSIX spec equivalent.

> That would require auditing all uses of ISO C APIs within any POSIX code brought into the Zephyr environment to ensure they do not rely on any POSIX extensions to the ISO C specification.

If you feel like doing that, I guess go for it?

Patches maybe welcome?

> Given that POSIX specifications are not supposed to conflict with ISO C specifications, but only to refine them, it seems like Zephyr should require a POSIX conforming C library when the application requests POSIX functionality?

Maybe. I'm ok with just using the ISO C spec for C standards though.

> Or at least make it clear to the application that they are operating outside known behavior when mixing non-POSIX implementations of C library functions with POSIX functionality.

Maybe. Again, I'm ok with just using the ISO C spec, and I think that's what the POSIX spec says to do too.

I can understand your concern though as a libc maintainer.

File: doc/services/portability/posix/conformance/index.rst [hunk 4]
```diff
@@ -61,7 +68,7 @@ POSIX System Interfaces
     _POSIX_REALTIME_SIGNALS, -1, :ref:`†<posix_undefined_behaviour>`
     :ref:`_POSIX_SEMAPHORES<posix_option_group_semaphores>`, 200809L, :kconfig:option:`CONFIG_PTHREAD_IPC`
     :ref:`_POSIX_SPIN_LOCKS<posix_option_group_spin_locks>`, 200809L, :kconfig:option:`CONFIG_PTHREAD_SPINLOCK`
-    _POSIX_THREAD_SAFE_FUNCTIONS, 200809L,
+    :ref:`_POSIX_THREAD_SAFE_FUNCTIONS<posix_thread_safe_functions>`, 200809L,
```
Comment (cfriedt) [inline doc/services/portability/posix/conformance/index.rst:71]: Just noticed - this should be `-1` since there are gaps in the table.

Comment (cfriedt) [reply doc/services/portability/posix/conformance/index.rst]: Fixed & rebased

File: doc/services/portability/posix/option_groups/index.rst (hunk 3 above)
Comment (cfriedt) [reply doc/services/portability/posix/option_groups/index.rst:128]: @keith-packard - I'm not sure if there is any further clarification I can attempt to make here outside of what is already specified by IEEE.

Feel free to unresolve the conversation if you feel it's necessary. However, I think it's well outside the scope of this PR.

**Comment by cfriedt** [comment]
Marked `_POSIX_THREAD_SAFE_FUNCTIONS` as `-1`, since it became obvious that there are some pieces missing still

**Comment by cfriedt** [comment]
@kartben, @ycsin, @keith-packard  - any further changes required?

**Review by kartben — APPROVED** [review]

**Review by ycsin — APPROVED** [review]
